### PR TITLE
Final review

### PR DESCRIPTION
**Pull request by ishranga11**
- Added complete dao module
- Added complete models module
- Corrected MySQLConfig, GScribeApplication, GScribeConfiguration, gscribe.yaml to take mysql credentials from YAML file.

### CODE REVIEW

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java [hunk 1]
```diff
@@ -16,5 +16,79 @@
...
+     * Queries exam metadata of a particular exam identified by examID and created by user identified by userID
+     * The userID is a unique user id generated by google for the user
+     * Combination of examID and userID is used to query to avoid querying exam created by another user
+     *
+     * @param examID ( to identify particular exam )
+     * @param userID ( unique user id of user, here paper setter )
+     * @return exam metadata object
+     */
+    @Mapper(ExamMetadataMapper.class)
+    @SqlQuery("SELECT * from exam where id = :exam_id and created_by=:userId")
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/ExamMetadataDao.java:41]: Add email.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            try {
+                Question question = objectMapper.readValue(resultSet.getString("question"), Question.class);
+                if (question.getType().equals(QuestionType.MCQ)) {
+                    return objectMapper.readValue(resultSet.getString("question"), MultipleChoiceQuestion.class);
+                } else if (question.getType().equals(QuestionType.SUBJECTIVE)) {
+                    return objectMapper.readValue(resultSet.getString("question"), SubjectiveQuestion.class);
+                }
+                return null;
+            } catch (JsonProcessingException e) {
+                e.printStackTrace();
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java:76]: Use logger.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                return null;
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java]: Don't catch it here, instead bubble it up like SQLException.

File: appengine/src/main/java/com/google/googleinterns/gscribe/models/ExamMetadata.java [hunk 4]
```diff
@@ -16,5 +16,90 @@
...
+    // TODO: add version for exams
+    private String spreadsheetID;
+    private String sheetName;
+    private String userID;
+    private int id;
+    private int duration;
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSS")
+    private Timestamp createdOn;
+
+    public ExamMetadata(String spreadsheetID, String sheetName, String userID, int id, int duration, Timestamp createdOn) {
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/models/ExamMetadata.java:37]: Use builder pattern, you can autogenerate builder in IntelliJ.

File: appengine/src/test/resources/schema.sql [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  end_time timestamp NULL DEFAULT NULL,
+  primary key (id),
+  KEY examtakenDB_to_examDB_examID_idx (exam_id),
+  CONSTRAINT examinstanceDB_to_examDB_exam_id FOREIGN KEY (exam_id) REFERENCES exam (id)
+);
+
+DROP TABLE IF EXISTS answers;
+create table answers (
+  exam_instance_id int not null,
+  answers varchar(1000) not null,
```
Comment (dhruvpancholi) [inline appengine/src/test/resources/schema.sql]: Increase the size for both questions and answers, and maybe use text field.

File: appengine/src/test/java/com/google/googleinterns/gscribe/provider/DBIProvider.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.InputStream;
+import java.io.StringWriter;
+
+public class DBIProvider {
+
+    public DBI getDBI() throws IOException {
+
+        DBI dbi = new DBI("jdbc:h2:mem:test;DB_CLOSE_DELAY=-1");
+        InputStream in = QuestionsDaoTest.class.getResourceAsStream("/schema.sql");
+        System.out.println(in.toString());
```
Comment (dhruvpancholi) [inline appengine/src/test/java/com/google/googleinterns/gscribe/provider/DBIProvider.java]: Remove all println.

File: appengine/src/test/java/module/DBIProvider.java [hunk 7]
```diff
@@ -14,10 +14,19 @@
...
 
-package com.google.googleinterns.gscribe.resources.io.exception;
+package module;
 
-public class AuthenticationException extends RuntimeException {
-    public AuthenticationException(String message) {
-        super(message);
+import org.skife.jdbi.v2.DBI;
+
+public class DBIProvider {
```
Comment (dhruvpancholi) [inline appengine/src/test/java/module/DBIProvider.java]: Is this required? There are two DBIProvider in src/test

File: appengine/src/main/java/com/google/googleinterns/gscribe/modules/DaoModule.java [hunk 8]
```diff
@@ -40,4 +40,27 @@ public ExamMetadataDao examMetadataDaoProvider(DBI dbi) {
         return dbi.onDemand(ExamMetadataDao.class);
     }
 
+    @Inject
+    @Provides
+    @Singleton
+    public QuestionsDao examDaoProvider(DBI dbi, ObjectMapper objectMapper) {
+        dbi.registerMapper(new QuestionsDao.QuestionsMapper(objectMapper));
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/modules/DaoModule.java]: You can inject the QuestionMapper as well, just have an Inject constructor.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    class AnswersMapper implements ResultSetMapper<Answers> {
+
+        ObjectMapper objectMapper;
+
+        public AnswersMapper(ObjectMapper objectMapper) {
+            this.objectMapper = objectMapper;
+        }
+
+        @Override
+        public Answers map(int i, ResultSet resultSet, StatementContext statementContext) throws SQLException {
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:62]: throws SQLException, JsonProcessingException

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            try {
+                return objectMapper.readValue(resultSet.getString("answers"), Answers.class);
+            } catch (JsonProcessingException e) {
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:65]: Remove catch

**Review by dhruvpancholi — CHANGES_REQUESTED** [review]

File: appengine/src/test/resources/schema.sql (hunk 5 above)
Comment (ishranga11) [reply appengine/src/test/resources/schema.sql]: As the text field in the in-memory database is getting stored as s clob field and giving a different output as expected so I thought of using varchar only. I will change the limit to 5000.

File: appengine/src/test/java/com/google/googleinterns/gscribe/provider/DBIProvider.java (hunk 6 above)
Comment (ishranga11) [reply appengine/src/test/java/com/google/googleinterns/gscribe/provider/DBIProvider.java]: Forgot to remove while uploading. Removed now.

File: appengine/src/test/java/module/DBIProvider.java (hunk 7 above)
Comment (ishranga11) [reply appengine/src/test/java/module/DBIProvider.java]: Forgot to remove while changing branches. This provider will be deleted.

File: appengine/src/main/java/com/google/googleinterns/gscribe/modules/DaoModule.java (hunk 8 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/modules/DaoModule.java]: Added a new MapperModule for this.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java (hunk 9 above)
Comment (ishranga11) [reply appengine/src/main/java/com/google/googleinterns/gscribe/dao/AnswerDao.java:62]: As the function Overrides the map function of ResultSetMapper, only SQLException can be thrown due to the function structure.

File: appengine/src/main/java/com/google/googleinterns/gscribe/modules/MapperModule.java [hunk 11]
```diff
@@ -16,27 +16,28 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.googleinterns.gscribe.dao.AnswerDao;
+import com.google.googleinterns.gscribe.dao.QuestionsDao;
 import com.google.inject.AbstractModule;
+import com.google.inject.Inject;
 import com.google.inject.Provides;
 import com.google.inject.Singleton;
 
-public class ConfigModule extends AbstractModule {
+public class MapperModule extends AbstractModule {
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/modules/MapperModule.java]: This is not required, you can directly create a constructor and inject ObjectMapper into it. It should automatically pickup.

File: appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                for (JsonNode question : questions) {
+                    if (question.get("type").asText().equals("MCQ")) {
+                        questionsList.getQuestions().add(objectMapper.treeToValue(question, MultipleChoiceQuestion.class));
+                    } else if (question.get("type").asText().equals("SUBJECTIVE")) {
+                        questionsList.getQuestions().add(objectMapper.treeToValue(question, SubjectiveQuestion.class));
+                    }
+                }
+                return questionsList;
+            } catch (JsonProcessingException e) {
+                System.out.println(e.getMessage());
```
Comment (dhruvpancholi) [inline appengine/src/main/java/com/google/googleinterns/gscribe/dao/QuestionsDao.java]: Remove this, or use logger.

**Review by dhruvpancholi — APPROVED** [review]
LGTM

**Pull request merged**